Source/MLXNN/Pooling.swift

```diff
@@ -0,0 +1,104 @@
+// Copyright © 2024 Apple Inc.
+
+import Foundation
+import MLX
+
+public class Pool: Module, UnaryLayer {
+
+    let kernelSize: [Int]
+    let stride: [Int]
+    let axes: [Int]
+    let poolingOp: (MLXArray, [Int]) -> MLXArray
+
+    init(
+        kernelSize: [Int],
+        stride: [Int],
+        poolingOp: @escaping (MLXArray, [Int]) -> MLXArray
+    ) {
+        self.kernelSize = kernelSize
+        self.stride = stride
+        self.axes = Array((-1 * kernelSize.count - 1) ..< -1)
+        print(axes)
+        self.poolingOp = poolingOp
+    }
+
+    public func callAsFunction(_ x: MLX.MLXArray) -> MLX.MLXArray {
+        let shape = x.shape
+
+        var finalShape = [shape[0]]
+        finalShape += zip(zip(shape.dropFirst().dropLast(), kernelSize), stride).map {
+            (t, stride) in
+            let (size, window) = t
+            return (size - window) / stride + 1
+        }
+        finalShape += kernelSize
+        finalShape += [shape.last!]
+
+        let strides = (shape + [1]).reversed().reduce([] as [Int]) { partial, a in
+            guard let element = partial.last else { return [a] }
+            return partial + [a * element]
+        }.reversed().dropFirst()
+        let middleStrides = strides.dropFirst().dropLast()
+        var finalStrides: [Int] = [strides.first!]
+        finalStrides += zip(middleStrides, stride).map { $0 * $1 }
+        finalStrides += middleStrides
+        finalStrides += [1]
+
+        let strided = asStrided(x, finalShape, strides: finalStrides)
+        return poolingOp(strided, axes)
+    }
+}
+
+/// Applies 2-dimensional max pooling.
+public class MaxPool2d: Pool {
+
+    /// Applies 2-dimensional max pooling.
+    ///
+    /// The input is expected to be `NHWC`
```

Comment:
Might be useful to include information on the output tensor: 

- https://github.com/ml-explore/mlx/blob/7dcdd88e27ae1f3e3fe2a4f91239722e7a4fe196/python/mlx/nn/layers/pooling.py#L216

We don't have a way to represent the math (in docc).  In some cases I used a link to the python docs for more information, if I thought it was important.